lib/concurrent/synchronization.rb

```diff
@@ -6,23 +6,26 @@
 require 'concurrent/synchronization/rbx_object'
 
 module Concurrent
+  # {include:file:doc/synchronization.md}
   module Synchronization
-    class Object < case
-                           when Concurrent.on_jruby?
-                             JavaObject
+    Implementation = case
+                     when Concurrent.on_jruby?
+                       JavaObject
+                     when Concurrent.on_cruby? && (RUBY_VERSION.split('.').map(&:to_i) <=> [1, 9, 3]) <= 0
+                       MonitorObject
+                     when Concurrent.on_cruby?
+                       MutexObject
+                     when Concurrent.on_rbx?
+                       RbxObject
+                     else
+                       MutexObject
+                     end
+    private_constant :Implementation
 
-                           when Concurrent.on_cruby? && (RUBY_VERSION.split('.').map(&:to_i) <=> [1, 9, 3]) <= 0
-                             MonitorObject
-
-                           when Concurrent.on_cruby?
-                             MutexObject
-
-                           when Concurrent.on_rbx?
-                             RbxObject
-
-                           else
-                             MutexObject
-                           end
+    # @see AbstractObject
+    class Object < Implementation
     end
   end
 end
+
+require 'concurrent/synchronization/immutable_struct'
```

Comment:
Made it little better in 4993b2c. It's required here following convention from `atomics.rb` file and similar, but I do not feel strongly about this one way or the other.